Wikipedia/Code/WMFArticleViewController.m

```diff
@@ -97,7 +97,11 @@ - (nullable UIActivityViewController *)sharingActivityViewControllerWithTextSnip
     NSURL *url = [NSURL wmf_desktopURLForURL:self.url];
 
     if (url) {
-        [items addObject:[[NSURL alloc] initWithString:[NSString stringWithFormat:@"%@?%@", url.absoluteString, @"wprov=sfsi1"]]];
+        NSString *urlAsString = url.absoluteString;
+        if([urlAsString hasSuffix:@"!"]) {
+            urlAsString = [urlAsString stringByReplacingOccurrencesOfString:@"!" withString:@"%21"];
+        }
+        [items addObject:[[NSURL alloc] initWithString:[NSString stringWithFormat:@"%@?%@", urlAsString, @"wprov=sfsi1"]]];
```

Comment:
Regardless of whether or not the `!` gets percent-encoded, the old code should be updated to use `NSURLComponents` and `NSURLQueryItem` to add `wprov=sfsi1` to the URL ([here's a similar use](https://github.com/wikimedia/wikipedia-ios/blob/7402cc2db9105910da1989c7dd4e0eca50eaad4e/Wikipedia/Code/WMFProxyServer.m#L260)). Also, instead of using a hardcoded `%21`, the built in `stringByAddingPercentEncodingWithAllowedCharacters:` could be used on the `path` property with a character set based on `[NSCharacterSet URLPathAllowedCharacterSet]` with `!` removed. That would then be set to `percentEncodedPath`.